tsfc/finatinterface.py

```diff
@@ -134,87 +128,119 @@ def convert_finiteelement(element, shape_innermost=True):
             lmbda = finat.GaussLegendre
         else:
             raise ValueError("Variant %r not supported on %s" % (kind, element.cell()))
-    return lmbda(cell, element.degree())
+    return lmbda(cell, element.degree()), set()
+
+
+# Element modifiers and compound element types
+@convert.register(ufl.BrokenElement)
+def convert_brokenelement(element, **kwargs):
+    finat_elem, deps = _create_element(element._element, **kwargs)
+    return finat.DiscontinuousElement(finat_elem), deps
 
 
-# EnrichedElement case
 @convert.register(ufl.EnrichedElement)
-def convert_enrichedelement(element, shape_innermost=True):
-    return finat.EnrichedElement([create_element(elem, shape_innermost)
-                                  for elem in element._elements])
+def convert_enrichedelement(element, **kwargs):
+    elements, deps = zip(*[_create_element(elem, **kwargs)
+                           for elem in element._elements])
+    return finat.EnrichedElement(elements), set.union(*deps)
 
 
-# Generic MixedElement case
 @convert.register(ufl.MixedElement)
-def convert_mixedelement(element, shape_innermost=True):
-    return finat.MixedElement([create_element(elem, shape_innermost)
-                               for elem in element.sub_elements()])
+def convert_mixedelement(element, **kwargs):
+    elements, deps = zip(*[_create_element(elem, **kwargs)
+                           for elem in element.sub_elements()])
+    return finat.MixedElement(elements), set.union(*deps)
 
 
-# VectorElement case
 @convert.register(ufl.VectorElement)
-def convert_vectorelement(element, shape_innermost=True):
-    scalar_element = create_element(element.sub_elements()[0], shape_innermost)
-    return finat.TensorFiniteElement(scalar_element,
-                                     (element.num_sub_elements(),),
-                                     transpose=not shape_innermost)
+def convert_vectorelement(element, **kwargs):
+    scalar_elem, deps = _create_element(element.sub_elements()[0], **kwargs)
+    shape = (element.num_sub_elements(),)
+    shape_innermost = kwargs["shape_innermost"]
+    return (finat.TensorFiniteElement(scalar_elem, shape, not shape_innermost),
+            deps | {"shape_innermost"})
 
 
-# TensorElement case
 @convert.register(ufl.TensorElement)
-def convert_tensorelement(element, shape_innermost=True):
-    scalar_element = create_element(element.sub_elements()[0], shape_innermost)
-    return finat.TensorFiniteElement(scalar_element,
-                                     element.reference_value_shape(),
-                                     transpose=not shape_innermost)
+def convert_tensorelement(element, **kwargs):
+    scalar_elem, deps = _create_element(element.sub_elements()[0], **kwargs)
+    shape = element.reference_value_shape()
+    shape_innermost = kwargs["shape_innermost"]
+    return (finat.TensorFiniteElement(scalar_elem, shape, not shape_innermost),
+            deps | {"shape_innermost"})
```

Comment:
OK, so `deps` records which `kwargs` affected element conversion, and then the cache key records `name, value` pairs.